eclair-core/src/main/scala/fr/acinq/eclair/channel/Channel.scala

```diff
@@ -1678,16 +1678,15 @@ class Channel(val nodeParams: NodeParams, val wallet: EclairWallet, remoteNodeId
         cancelTimer(RevocationTimeout.toString)
       }
 
-      // if channel is private, we send the channel_update directly to remote
-      // they need it "to learn the other end's forwarding parameters" (BOLT 7)
-      (stateData, nextStateData) match {
-        case (d1: DATA_NORMAL, d2: DATA_NORMAL) if !d1.commitments.announceChannel && !d1.buried && d2.buried =>
-          // for a private channel, when the tx was just buried we need to send the channel_update to our peer (even if it didn't change)
+      (state, nextState, stateData, nextStateData) match {
+        // if channel is private, we send the channel_update directly to remote on each reconnect
+        // they need it "to learn the other end's forwarding parameters" (BOLT 7)
+        case (NORMAL, NORMAL, d1: DATA_NORMAL, d2: DATA_NORMAL) if !d1.commitments.announceChannel && (!d1.buried && d2.buried || d1.channelUpdate != d2.channelUpdate) =>
```

Comment:
I think that's incorrect. You still always want to wait for `d2.buried` before sending a `channel_update` (which isn't the case here if `d1.channelUpdate != d2.channelUpdate`.

I'm a bit weary of matching now on states + data, there may be a case we're losing compared to the previous code.

Here is my take:

```scala
// if channel is private, we send the channel_update directly to remote
// they need it "to learn the other end's forwarding parameters" (BOLT 7)
(state, nextState, stateData, nextStateData) match {
  case (_, _, d1: DATA_NORMAL, d2: DATA_NORMAL) if !d1.commitments.announceChannel && !d1.buried && d2.buried =>
    // for a private channel, when the tx was just buried we need to send the channel_update to our peer (even if it didn't change)
    forwarder ! d2.channelUpdate
  case (OFFLINE, NORMAL, d1: DATA_NORMAL, d2: DATA_NORMAL) if !d1.commitments.announceChannel && d2.buried =>
    // otherwise if we're coming back online, we rebroadcast the latest channel_update
    // this makes sure that if the channel_update was missed, we have a chance to re-send it
    forwarder ! d2.channelUpdate
  case (_, _, d1: DATA_NORMAL, d2: DATA_NORMAL) if !d1.commitments.announceChannel && d1.channelUpdate != d2.channelUpdate && d2.buried =>
    // otherwise, we only send it when it is different, and tx is already buried
    forwarder ! d2.channelUpdate
  case _ => ()
}
```

I believe it's safer as it keeps the existing behavior and only adds a new case when coming back online. WDYT?

Reply:
> I think that's incorrect. 

It's less code and not sure it's strictly incorrect: `NORMAL_DATA` is present so at least default 3 confirmations are there, occasional premature update at this point won't add too much risk.

But I'm also totally fine with using your code.